### DISCUSSION TITLE
DfMp3_Error_RxTimeout on any get method

### DISCUSSION THREAD

**Discussion by PowercoderJr** [Makuna/DFMiniMp3#82] [Q&A] [answered]
I use arduino nano and try to get, for example, initial volume. The code is as follows:

```
SoftwareSerial secondarySerial(10, 11);
DFMiniMp3<SoftwareSerial, Mp3Notify> mp3(secondarySerial);

void setup() {
    Serial.begin(9600);
    Serial.println("Initializing...");
    mp3.begin();
    mp3.reset();
    Serial.println(String("Initial volume = ") + mp3.getVolume());
}
```

Output:
> 02:02:56:730 -> Initializing...
> 02:03:07:444 -> 
> Com Error 129
> Initial volume = 0

The only mention of this error I've found is #19 where is said that it is likely wiring problem. I've made connections as shown by [link in readme](https://wiki.dfrobot.com/DFPlayer_Mini_SKU_DFR0299). I've tried different arduino pins for `SoftwareSerial` and different DFMP3 modules.

Reference:

![reference](https://user-images.githubusercontent.com/25909264/161964806-94c720e4-a2ab-4011-bf22-74560bb9b1fb.png)

My wiring:

![wiring1](https://user-images.githubusercontent.com/25909264/161963198-7b07c96a-3e65-4e19-aa80-c82badb8813e.png)
![wiring2](https://user-images.githubusercontent.com/25909264/161963280-84a478be-4c10-4a91-9562-795ce169faae.png)

**Comment by Makuna** [comment]
This is more of a question than a bug; and should have been done in the discussions area.  I moved it there.

**Comment by Makuna** [comment]
That error is `DfMp3_Error_RxTimeout`, the module didn't respond/send any data.  This usually is due to the pins not connected correctly; but on rare cases due to an API not being supported (the chips are not very standardized unfortunately).

Did you try one of the examples?
Did you try swapping the pins? Some modules will label RX and mean connect this to your Arduino's RX pin; some will label meaning it's the modules receive pin.
I noticed you added a resistor on a data line.  Did you try removing it?

**Reply by PowercoderJr** [reply]
I started from PlayRandom and after getting the error came to this minimal code. It actually played random sounds, playing a specific sound from a specific folder worked too but setting volume didn't.

I tried to swap pins and use other pins of arduino. I tried 3 different dfplayers (which are the same though).

I tried this scheme without resistor.

**Reply by PowercoderJr** [reply]
By the way, are these examples for current release? My IDE says `Mp3Notify` class methods doesn't need `DfMp3& mp3` arguments.

**Reply by Makuna** [reply]
> By the way, are these examples for current release? My IDE says `Mp3Notify` class methods doesn't need `DfMp3& mp3` arguments.

This "message" is due to your specific (or the examples) implementation of the methods in the notify class are not using the passed in mp3 AND that you have warnings set to ALL.  There are several ways to suppress this message.  See https://stackoverflow.com/questions/1486904/how-do-i-best-silence-a-warning-about-unused-variables.  
This compiles for so many different compiler versions, and only one of the latest did they add the `[[maybe_unused]]`.

**Reply by PowercoderJr** [reply]
> This "message" is due to your specific (or the examples) implementation of the methods in the notify class are not using the passed in mp3 AND that you have warnings set to ALL. There are several ways to suppress this message. See https://stackoverflow.com/questions/1486904/how-do-i-best-silence-a-warning-about-unused-variables. This compiles for so many different compiler versions, and only one of the latest did they add the `[[maybe_unused]]`.

I took [the example from repo](https://github.com/Makuna/DFMiniMp3/blob/master/examples/PlayRandom/PlayRandom.ino) and oh, I see the repo source code is different from the one I got from library manager.

UPD: reinstalling library from github source let me run the example as is without compiling errors but it didn't solve the main problem.

**Comment by Makuna** [comment]
Something to try, add a `delay(10000)` after the reset.  Maybe your module takes a long time to recover from the reset.
Another thing to try, don't use the nano 5v output to power the module.  While some Arduinos have plenty of power on their 5v rail, the nano's tend not too.  But in general, its best to always power other modules from a separate and known power rail with known amperage ratings.  Of course, keep all the GNDs connected.
And then, try removing the `reset()` call altogether.

**Reply by PowercoderJr** [reply]
Sadly, not delay neither removing reset worked. I'll try to power up with separate power soon.

**Reply by Makuna** [reply]
Are you comfortable modifying my code (on your machine locally) to add some debug output?  I can hand you some lines to add and it will help shed some light.

**Reply by PowercoderJr** [reply]
> Are you comfortable modifying my code (on your machine locally) to add some debug output? I can hand you some lines to add and it will help shed some light.

Yes, please

**Reply by PowercoderJr** [reply]
> Something to try, add a `delay(10000)` after the reset. Maybe your module takes a long time to recover from the reset. Another thing to try, don't use the nano 5v output to power the module. While some Arduinos have plenty of power on their 5v rail, the nano's tend not too. But in general, its best to always power other modules from a separate and known power rail with known amperage ratings. Of course, keep all the GNDs connected. And then, try removing the `reset()` call altogether.

Having a sparate power supply I managed to get this output couple of times:

> 
> 01:03:08:966 -> initializing...
> 01:03:19:077 -> SD Card, online
> 01:03:19:185 -> volume 30
> 01:03:29:248 -> 
> Com Error 129
> files 0
> 01:03:29:358 -> playback mode 5
> starting...
> 01:03:30:672 -> Play finished for # 3
> 01:03:31:547 -> Play finished for # 2
> 01:03:32:216 -> Play finished for # 3
> ...

`setVolume()` also worked. But it is still unstable. For each rerun I can get success or not in getting initial volume or files count independently. 
I use 2x1.5V batteries with [a transformer](https://aliexpress.ru/item/32820218683.html). Do I need even more current?

**Reply by PowercoderJr** [reply]
I noticed I have constant 3.5V on player's TX, is it ok?

**Reply by Makuna** [reply]
Ok, this sounds like a power supply issue.  The unit really wants a 3.3v power input, its tolerant of 5v, but a solid 3.3v is best.

TTL Serial is high when idle so that is normal'ish, (https://www.sparkfun.com/tutorials/215#:~:text=UARTs%20transmit%20one%20bit%20at%20a%20time%20at,and%20Vcc%2C%20which%20is%20often%205V%20or%203.3V.), but the unit is a 3.3v based (input tolerance upto 5v) so 3.5v seems a little high.
Note, that 1K resistor is needed (MCU TX -> 1K -> Mp3), if you haven't put it back.

What type of batteries are you using for those two?  Alkaline or Nmh rechargables?  AA or AAA?

Alkaline while provides a true 1.5v, they often can't source much power draw.  Nmh only provides about 1.25v but can source more power draw at once.

When supplying that boost converter with 2-3v, it will provide a max power of 380ma.  It doesn't state what its nominal power output would be.  I would add another cell if you can (3x1.5v/1.25v).  Even better, use a power supply instead of batteries (wall wart/etc).  I always suggest using a DC Buck converter with a higher voltage plug in supply (if you don't have a bench supply) at least for bench work. Boost converters like you have just can't provide the power you may need.

**Reply by PowercoderJr** [reply]
@Makuna, I use AAA alkaline batteries. I want to put my scheme into an existing body so I'm restricted to use 2xAAA and trying to keep it as small as possible. But still thank you for your help. Datasheet says its working voltage is 3.2~5.0V and my batteries give ~2.9 in sum, so I decided to use transformer.

**Reply by PowercoderJr** [reply]
> It doesn't state what its nominal power output would be

Oh wow, it is only 120 mA. How much does player need? Can't find the value in the datasheet.

> I would add another cell if you can (3x1.5v/1.25v)

I have added one and removed the transformer. Didn't work - same result as with 2xAAA + trnsformer.

> `setVolume()` also worked

It seems like it actually didn't but the volume decreased due to lower power. The only stable thing is source action prints ("SD Card, online", "SD Card, removed").

**Reply by Makuna** [reply]
Something you may want to consider, run your Arduino at 3.3v.  AVR chips can run at 3.3v and will use less power; but maybe not be able to still clock at 16mhz (check the spec sheet on the CPU on your Arduino).   They use to make Pro versions on your Arduino that did run at 3.3v (I have a few).

**Comment by PowercoderJr** [comment, marked as answer]
I finally made it work! @Makuna, thank you a lot for your attention, it helped.
The problem was all in power. It was not enough to use power of USB connection even though the player had separate power supply. I powered both arduino and the player from 3x1.5 V AAA and it works fine and stable now. Woohoo! And there is a micromoment that between `setVolume()` and `getVolume()` some delay is needed. I didn't seek for an exact value but 100 ms is enough.

UPD: 2x1.5 V AAA and a transformer work fine too.

**Reply by Makuna** [reply]
What happens without the delay(100) between those two?  Do you get the timeout error?  Bad value? 

There are a few commands that require a little space after them before another command can be issued.  I am wondering if the newer MP3 modules now have the setVolume neeeding this.  Try setVolume and then something else and see it if causes the same issue.  If so, I can add the required spacing (part of the internal send) by changing....
```
    void setVolume(uint8_t volume)
    {
        sendPacket(0x06, volume);
    }
```
to ...
```
    void setVolume(uint8_t volume)
    {
        sendPacket(0x06, volume, 100);
    }
```

**Reply by PowercoderJr** [reply]
It is timeout, 129, and yes, it happens with `getTotalTrackCount()` and `getPlaybackMode()` after `setVolume()` without delay. Looks like space is good here.
I also noticed a 500 ms space more needed after `reset()` in my case. Existing 600 ms space + additional 300 ms delay is unstable. Space 1100 works fine,

**Reply by Makuna** [reply]
https://github.com/Makuna/DFMiniMp3/pull/84
if you find any others let me know or submit changes
